client/windows/CMapOverview.cpp

```diff
@@ -0,0 +1,240 @@
+				// if object at tile is owned - it will be colored as its owner
+				for (const CGObjectInstance *obj : tile.blockingObjects)
+				{
+					PlayerColor player = obj->getOwner();
+					if(player == PlayerColor::NEUTRAL)
+					{
+						color = graphics->neutralColor;
+						break;
+					}
+					if (player.isValidPlayer())
+					{
+						color = graphics->playerColors[player.getNum()];
+						break;
+					}
+				}
+
+			canvas.drawPoint(Point(x, y), color);
+		}
+	
+	return canvas;
+}
+
+std::vector<std::shared_ptr<IImage>> CMapOverview::CMapOverviewWidget::createMinimaps(ResourcePath resource, Point size) const
+{
+	std::vector<std::shared_ptr<IImage>> ret = std::vector<std::shared_ptr<IImage>>();
+
+	CMapService mapService;
+	std::unique_ptr<CMap> map;
+	try
+	{
+		map = mapService.loadMap(resource);
+	}
+	catch (...)
+	{
+		return ret;
+	}
+
+	return createMinimaps(map, size);
+}
+
+std::vector<std::shared_ptr<IImage>> CMapOverview::CMapOverviewWidget::createMinimaps(std::unique_ptr<CMap> & map, Point size) const
+{
+	std::vector<std::shared_ptr<IImage>> ret = std::vector<std::shared_ptr<IImage>>();
+
+	for(int i = 0; i < (map->twoLevel ? 2 : 1); i++)
+	{
+		Canvas canvas = createMinimapForLayer(map, i);
+		Canvas canvasScaled = Canvas(size);
+		canvasScaled.drawScaled(canvas, Point(0, 0), size);
+		std::shared_ptr<IImage> img = GH.renderHandler().createImage(canvasScaled.getInternalSurface());
+		
+		ret.push_back(img);
+	}
+
+	return ret;
+}
+
+std::shared_ptr<TransparentFilledRectangle> CMapOverview::CMapOverviewWidget::CMapOverviewWidget::buildDrawTransparentRect(const JsonNode & config) const
+{
+	logGlobal->debug("Building widget drawTransparentRect");
+
+	auto rect = readRect(config["rect"]);
+	auto color = readColor(config["color"]);
+	if(!config["colorLine"].isNull())
+	{
+		auto colorLine = readColor(config["colorLine"]);
+		return std::make_shared<TransparentFilledRectangle>(rect, color, colorLine);
+	}
+	return std::make_shared<TransparentFilledRectangle>(rect, color);
+}
+
+std::shared_ptr<CPicture> CMapOverview::CMapOverviewWidget::buildDrawMinimap(const JsonNode & config) const
+{
+	logGlobal->debug("Building widget drawMinimap");
+
+	auto rect = readRect(config["rect"]);
+	auto id = config["id"].Integer();
+
+	if(!renderImage)
+		return nullptr;
+
+	ResourcePath res = ResourcePath(parent.resource.getName(), EResType::MAP);
+	std::unique_ptr<CMap> campaignMap = nullptr;
+	if(parent.tabType != ESelectionScreen::newGame)
+	{
+		CLoadFile lf(*CResourceHandler::get()->getResourceName(ResourcePath(parent.resource.getName(), EResType::SAVEGAME)), MINIMAL_SERIALIZATION_VERSION);
+		lf.checkMagicBytes(SAVEGAME_MAGIC);
+
+		std::unique_ptr<CMapHeader> mapHeader = std::make_unique<CMapHeader>();
+		StartInfo * startInfo;
+		lf >> *(mapHeader) >> startInfo;
+
+		if(startInfo->campState)
+			campaignMap = startInfo->campState->getMap(*startInfo->campState->currentScenario());
+		res = ResourcePath(startInfo->fileURI, EResType::MAP);
```

Comment:
what will happen if save game was sent between players and other player doesn't have campaign installed (or removed it locally)?

Reply:
Same as if you only have save and no map: no preview.